File: java/src/main/java/benchmarks/Maze.java
package benchmarks;

import java.util.*;

class MazeCellKind {
    public static final int WALL = 0;
    public static final int SPACE = 1;
    public static final int START = 2;
    public static final int FINISH = 3;
    public static final int BORDER = 4;
    public static final int PATH = 5;
}

class MazeCell {
    public int kind;
    public final MazeCell[] neighbors = new MazeCell[4];
    public int neighborCount;
    public final int x, y;

    public MazeCell(int x, int y) {
        this.kind = MazeCellKind.WALL;
        this.x = x;
        this.y = y;
        this.neighborCount = 0;
    }

    public void addNeighbor(MazeCell cell) {
        neighbors[neighborCount++] = cell;
    }

    public boolean isWalkable() {
        return kind == MazeCellKind.SPACE ||
               kind == MazeCellKind.START ||
               kind == MazeCellKind.FINISH;
    }

    public void reset() {
        if (kind == MazeCellKind.SPACE) {
            kind = MazeCellKind.WALL;
        }
    }
}

class Maze {
    private final int width, height;
    public final MazeCell[][] cells;
    private final MazeCell start, finish;
    private final Random random = new Random();

    public Maze(int w, int h) {
        this.width = w;
        this.height = h;
        this.cells = new MazeCell[height][width];

        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                cells[y][x] = new MazeCell(x, y);
            }
        }

        start = cells[1][1];
        finish = cells[height-2][width-2];
        start.kind = MazeCellKind.START;
        finish.kind = MazeCellKind.FINISH;

        updateNeighbors();
    }

    public void updateNeighbors() {
        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                MazeCell cell = cells[y][x];
                cell.neighborCount = 0;

                if (x > 0 && y > 0 && x < width - 1 && y < height - 1) {
                    cell.addNeighbor(cells[y-1][x]);
                    cell.addNeighbor(cells[y+1][x]);
                    cell.addNeighbor(cells[y][x+1]);
                    cell.addNeighbor(cells[y][x-1]);

                    for (int t = 0; t < 4; t++) {
                        int i = Helper.nextInt(4);
                        int j = Helper.nextInt(4);
                        if (i != j) {
                            MazeCell temp = cell.neighbors[i];
                            cell.neighbors[i] = cell.neighbors[j];
                            cell.neighbors[j] = temp;
                        }
                    }
                } else {
                    cell.kind = MazeCellKind.BORDER;
                }
            }
        }
    }

    public void reset() {
        for (MazeCell[] row : cells) {
            for (MazeCell cell : row) {
                cell.reset();
            }
        }
        start.kind = MazeCellKind.START;
        finish.kind = MazeCellKind.FINISH;
    }

    public void dig(MazeCell startCell) {
        Deque<MazeCell> stack = new ArrayDeque<>();
        stack.push(startCell);

        while (!stack.isEmpty()) {
            MazeCell cell = stack.pop();

            int walkable = 0;
            for (int i = 0; i < cell.neighborCount; i++) {
                if (cell.neighbors[i].isWalkable()) walkable++;
            }

            if (walkable != 1) continue;

            cell.kind = MazeCellKind.SPACE;

            for (int i = 0; i < cell.neighborCount; i++) {
                MazeCell n = cell.neighbors[i];
                if (n.kind == MazeCellKind.WALL) {
                    stack.push(n);
                }
            }
        }
    }

    public void ensureOpenFinish(MazeCell startCell) {
        Deque<MazeCell> stack = new ArrayDeque<>();
        stack.push(startCell);

        while (!stack.isEmpty()) {
            MazeCell cell = stack.pop();

            cell.kind = MazeCellKind.SPACE;

            int walkable = 0;
            for (int i = 0; i < cell.neighborCount; i++) {
                if (cell.neighbors[i].isWalkable()) walkable++;
            }

            if (walkable > 1) continue;

            for (int i = 0; i < cell.neighborCount; i++) {
                MazeCell n = cell.neighbors[i];
                if (n.kind == MazeCellKind.WALL) {
                    stack.push(n);
                }
            }
        }
    }

    public void generate() {
        for (int i = 0; i < start.neighborCount; i++) {
            MazeCell n = start.neighbors[i];
            if (n.kind == MazeCellKind.WALL) {
                dig(n);
            }
        }

        for (int i = 0; i < finish.neighborCount; i++) {
            MazeCell n = finish.neighbors[i];
            if (n.kind == MazeCellKind.WALL) {
                ensureOpenFinish(n);
            }
        }
    }

    public MazeCell middleCell() {
        return cells[height/2][width/2];
    }

    public MazeCell getStart() {
        return start;
    }
    public MazeCell getFinish() {
        return finish;
    }

    public long checksum() {
        long hasher = 2166136261L;
        long prime = 16777619L;

        for (MazeCell[] row : cells) {
            for (MazeCell cell : row) {
                if (cell.kind == MazeCellKind.SPACE) {
                    long val = ((long)cell.x * cell.y) & 0xFFFFFFFFL;
                    hasher = ((hasher ^ val) * prime) & 0xFFFFFFFFL;
                }
            }
        }
        return hasher;
    }

    public void printToConsole() {
        for (MazeCell[] row : cells) {
            for (MazeCell cell : row) {
                switch (cell.kind) {
                case MazeCellKind.SPACE:
                    System.out.print(" ");
                    break;
                case MazeCellKind.WALL:
                    System.out.print("\u001B[34m#\u001B[0m");
                    break;
                case MazeCellKind.BORDER:
                    System.out.print("\u001B[31mO\u001B[0m");
                    break;
                case MazeCellKind.START:
                    System.out.print("\u001B[32m>\u001B[0m");
                    break;
                case MazeCellKind.FINISH:
                    System.out.print("\u001B[32m<\u001B[0m");
                    break;
                case MazeCellKind.PATH:
                    System.out.print("\u001B[33m.\u001B[0m");
                    break;
                }
            }
            System.out.println();
        }
        System.out.println();
    }
}

class MazeGenerator extends Benchmark {
    private long resultVal;
    private final int width, height;
    private final Maze maze;

    public MazeGenerator() {
        this.width = (int) configVal("w");
        this.height = (int) configVal("h");
        this.maze = new Maze(width, height);
        this.resultVal = 0L;
    }

    @Override
    public String name() {
        return "Maze::Generator";
    }

    @Override
    public void prepare() {}

    @Override
    public void run(int iterationId) {
        maze.reset();
        maze.generate();
        resultVal = (resultVal + maze.middleCell().kind) & 0xFFFFFFFFL;
    }

    @Override
    public long checksum() {
        return (resultVal + maze.checksum()) & 0xFFFFFFFFL;
    }
}

class MazeBFS extends Benchmark {
    private long resultVal;
    private final int width, height;
    private final Maze maze;
    private List<MazeCell> path = new ArrayList<>();

    public MazeBFS() {
        this.width = (int) configVal("w");
        this.height = (int) configVal("h");
        this.maze = new Maze(width, height);
        this.resultVal = 0L;
    }

    @Override
    public String name() {
        return "Maze::BFS";
    }

    @Override
    public void prepare() {
        maze.generate();
    }

    private static class PathNode {
        MazeCell cell;
        int parent;
        PathNode(MazeCell cell, int parent) {
            this.cell = cell;
            this.parent = parent;
        }
    }

    private List<MazeCell> bfs(MazeCell start, MazeCell target) {
        if (start == target) {
            return Arrays.asList(start);
        }

        Deque<Integer> queue = new ArrayDeque<>();
        boolean[][] visited = new boolean[height][width];
        List<PathNode> pathNodes = new ArrayList<>();

        visited[start.y][start.x] = true;
        pathNodes.add(new PathNode(start, -1));
        queue.add(0);

        while (!queue.isEmpty()) {
            int pathId = queue.poll();
            MazeCell cell = pathNodes.get(pathId).cell;

            for (int i = 0; i < cell.neighborCount; i++) {
                MazeCell neighbor = cell.neighbors[i];

                if (neighbor == target) {
                    List<MazeCell> result = new ArrayList<>();
                    result.add(target);
                    int current = pathId;
                    while (current >= 0) {
                        result.add(pathNodes.get(current).cell);
                        current = pathNodes.get(current).parent;
                    }
                    Collections.reverse(result);
                    return result;
                }

                if (neighbor.isWalkable() && !visited[neighbor.y][neighbor.x]) {
                    visited[neighbor.y][neighbor.x] = true;
                    pathNodes.add(new PathNode(neighbor, pathId));
                    queue.add(pathNodes.size() - 1);
                }
            }
        }

        return new ArrayList<>();
    }

    private long midCellChecksum(List<MazeCell> p) {
        if (p.isEmpty()) return 0;
        MazeCell cell = p.get(p.size() / 2);
        return ((long)cell.x * cell.y) & 0xFFFFFFFFL;
    }

    @Override
    public void run(int iterationId) {
        path = bfs(maze.getStart(), maze.getFinish());
        resultVal = (resultVal + path.size()) & 0xFFFFFFFFL;
    }

    @Override
    public long checksum() {
        return (resultVal + midCellChecksum(path)) & 0xFFFFFFFFL;
    }
}

class MazeAStar extends Benchmark {
    private static class PriorityQueueItem implements Comparable<PriorityQueueItem> {
        int priority;
        int vertex;

        PriorityQueueItem(int priority, int vertex) {
            this.priority = priority;
            this.vertex = vertex;
        }

        @Override
        public int compareTo(PriorityQueueItem other) {
            if (this.priority != other.priority)
                return Integer.compare(this.priority, other.priority);
            return Integer.compare(this.vertex, other.vertex);
        }
    }

    private long resultVal;
    private final int width, height;
    private final Maze maze;
    private List<MazeCell> path = new ArrayList<>();
    private final int[] bestF;

    public MazeAStar() {
        this.width = (int) configVal("w");
        this.height = (int) configVal("h");
        this.maze = new Maze(width, height);
        this.resultVal = 0L;
        this.bestF = new int[width * height];
    }

    @Override
    public String name() {
        return "Maze::AStar";
    }

    @Override
    public void prepare() {
        maze.generate();
        Arrays.fill(bestF, Integer.MAX_VALUE);
    }

    private int heuristic(MazeCell a, MazeCell b) {
        return Math.abs(a.x - b.x) + Math.abs(a.y - b.y);
    }

    private int idx(int y, int x) {
        return y * width + x;
    }

    private List<MazeCell> astar(MazeCell start, MazeCell target) {
        if (start == target) return Arrays.asList(start);

        int size = width * height;

        int[] cameFrom = new int[size];
        int[] gScore = new int[size];
        Arrays.fill(cameFrom, -1);
        Arrays.fill(gScore, Integer.MAX_VALUE);

        int startIdx = idx(start.y, start.x);
        int targetIdx = idx(target.y, target.x);

        PriorityQueue<PriorityQueueItem> openSet = new PriorityQueue<>();

        int[] bestF = new int[size];
        Arrays.fill(bestF, Integer.MAX_VALUE);

        gScore[startIdx] = 0;
        int fStart = heuristic(start, target);
        openSet.offer(new PriorityQueueItem(fStart, startIdx));
        bestF[startIdx] = fStart;

        while (!openSet.isEmpty()) {
            PriorityQueueItem item = openSet.poll();
            int currentIdx = item.vertex;

            if (item.priority != bestF[currentIdx]) continue;

            if (currentIdx == targetIdx) {
                List<MazeCell> result = new ArrayList<>();
                int cur = currentIdx;
                while (cur != -1) {
                    int y = cur / width;
                    int x = cur % width;
                    result.add(maze.cells[y][x]);
                    cur = cameFrom[cur];
                }
                Collections.reverse(result);
                return result;
            }

            int currentY = currentIdx / width;
            int currentX = currentIdx % width;
            MazeCell current = maze.cells[currentY][currentX];
            int currentG = gScore[currentIdx];

            for (int i = 0; i < current.neighborCount; i++) {
                MazeCell neighbor = current.neighbors[i];
                if (!neighbor.isWalkable()) continue;

                int neighborIdx = idx(neighbor.y, neighbor.x);
                int tentativeG = currentG + 1;

                if (tentativeG < gScore[neighborIdx]) {
                    cameFrom[neighborIdx] = currentIdx;
                    gScore[neighborIdx] = tentativeG;
                    int fNew = tentativeG + heuristic(neighbor, target);

                    if (fNew < bestF[neighborIdx]) {
                        bestF[neighborIdx] = fNew;
                        openSet.offer(new PriorityQueueItem(fNew, neighborIdx));
                    }
                }
            }
        }

        return new ArrayList<>();
    }

    private long midCellChecksum(List<MazeCell> p) {
        if (p.isEmpty()) return 0;
        MazeCell cell = p.get(p.size() / 2);
        return ((long)cell.x * cell.y) & 0xFFFFFFFFL;
    }

    @Override
    public void run(int iterationId) {
        path = astar(maze.getStart(), maze.getFinish());
        resultVal = (resultVal + path.size()) & 0xFFFFFFFFL;
    }

    @Override
    public long checksum() {
        return (resultVal + midCellChecksum(path)) & 0xFFFFFFFFL;
    }
}

File: scala/src/main/scala/benchmarks/Maze.scala
package benchmarks

import java.util.{ArrayDeque, Deque, PriorityQueue}
import scala.collection.mutable
import scala.math

object MazeTypes {
  object CellKind {
    val WALL = 0
    val SPACE = 1
    val START = 2
    val FINISH = 3
    val BORDER = 4
    val PATH = 5

    def isWalkable(kind: Int): Boolean =
      kind == SPACE || kind == START || kind == FINISH
  }

  class Cell(val x: Int, val y: Int) {
    var kind: Int = CellKind.WALL
    val neighbors = mutable.ArrayBuffer.empty[Cell]

    def addNeighbor(cell: Cell): Unit = neighbors += cell
    def isWalkable: Boolean = CellKind.isWalkable(kind)
    def reset(): Unit = if (kind == CellKind.SPACE) kind = CellKind.WALL
  }

  class Maze(val width: Int, val height: Int) {
    val w = math.max(width, 5)
    val h = math.max(height, 5)

    val cells: Array[Array[Cell]] = Array.ofDim[Cell](h, w)

    private var _start: Cell = _
    private var _finish: Cell = _

    def start: Cell = _start
    def finish: Cell = _finish

    for (y <- 0 until h) {
      for (x <- 0 until w) {
        cells(y)(x) = new Cell(x, y)
      }
    }

    _start = cells(1)(1)
    _finish = cells(h - 2)(w - 2)
    _start.kind = CellKind.START
    _finish.kind = CellKind.FINISH

    updateNeighbors()

    def updateNeighbors(): Unit = {
      for (y <- 0 until h) {
        for (x <- 0 until w) {
          val cell = cells(y)(x)
          cell.neighbors.clear()

          if (x > 0 && y > 0 && x < w - 1 && y < h - 1) {
            cell.addNeighbor(cells(y - 1)(x))
            cell.addNeighbor(cells(y + 1)(x))
            cell.addNeighbor(cells(y)(x + 1))
            cell.addNeighbor(cells(y)(x - 1))

            for (_ <- 0 until 4) {
              val i = Helper.nextInt(4)
              val j = Helper.nextInt(4)
              if (i != j) {
                val temp = cell.neighbors(i)
                cell.neighbors(i) = cell.neighbors(j)
                cell.neighbors(j) = temp
              }
            }
          } else {
            cell.kind = CellKind.BORDER
          }
        }
      }
    }

    def reset(): Unit = {
      for (row <- cells; cell <- row) cell.reset()
      _start.kind = CellKind.START
      _finish.kind = CellKind.FINISH
    }

    def dig(startCell: Cell): Unit = {
      import scala.collection.mutable.ArrayStack

      val stack = new ArrayStack[Cell](w * h)
      stack.push(startCell)

      while (stack.nonEmpty) {
        val cell = stack.pop()

        var walkable = 0
        for (n <- cell.neighbors) {
          if (n.isWalkable) walkable += 1
        }

        if (walkable == 1) {
          cell.kind = CellKind.SPACE
          for (n <- cell.neighbors) {
            if (n.kind == CellKind.WALL) {
              stack.push(n)
            }
          }
        }
      }
    }

    def ensureOpenFinish(startCell: Cell): Unit = {
      val stack = mutable.Stack[Cell]()
      stack.push(startCell)

      while (stack.nonEmpty) {
        val cell = stack.pop()
        cell.kind = CellKind.SPACE

        val walkable = cell.neighbors.count(_.isWalkable)
        if (walkable <= 1) {
          cell.neighbors.filter(_.kind == CellKind.WALL).foreach(stack.push)
        }
      }
    }

    def generate(): Unit = {
      _start.neighbors.filter(_.kind == CellKind.WALL).foreach(dig)
      _finish.neighbors.filter(_.kind == CellKind.WALL).foreach(ensureOpenFinish)
    }

    def middleCell(): Cell = cells(h / 2)(w / 2)

    def checksum(): Long = {
      var hasher = 2166136261L
      val prime = 16777619L

      for (y <- 0 until h; x <- 0 until w) {
        if (cells(y)(x).kind == CellKind.SPACE) {
          val value = (x * y).toLong & 0xffffffffL
          hasher = ((hasher ^ value) * prime) & 0xffffffffL
        }
      }
      hasher
    }

    def printToConsole(): Unit = {
      for (y <- 0 until h) {
        for (x <- 0 until w) {
          cells(y)(x).kind match {
            case CellKind.SPACE  => print(" ")
            case CellKind.WALL   => print("\u001b[34m#\u001b[0m")
            case CellKind.BORDER => print("\u001b[31mO\u001b[0m")
            case CellKind.START  => print("\u001b[32m>\u001b[0m")
            case CellKind.FINISH => print("\u001b[32m<\u001b[0m")
            case CellKind.PATH   => print("\u001b[33m.\u001b[0m")
          }
        }
        println()
      }
      println()
    }
  }
}

class MazeGenerator extends Benchmark {
  import MazeTypes._

  private var resultVal: Long = 0L
  private val width: Int = configVal("w").toInt
  private val height: Int = configVal("h").toInt
  private var maze: Maze = _

  override def name(): String = "Maze::Generator"

  override def prepare(): Unit = {
    maze = new Maze(width, height)
    resultVal = 0L
  }

  override def run(iterationId: Int): Unit = {
    maze.reset()
    maze.generate()
    resultVal = (resultVal + maze.middleCell().kind) & 0xffffffffL
  }

  override def checksum(): Long = (resultVal + maze.checksum()) & 0xffffffffL
}

class MazeBFS extends Benchmark {
  import MazeTypes._

  case class PathNode(cell: Cell, parent: Int)

  private var resultVal: Long = 0L
  private val width: Int = configVal("w").toInt
  private val height: Int = configVal("h").toInt
  private var maze: Maze = _
  private var path: List[Cell] = Nil

  override def name(): String = "Maze::BFS"

  override def prepare(): Unit = {
    maze = new Maze(width, height)
    maze.generate()
    resultVal = 0L
    path = Nil
  }

  def bfs(start: Cell, target: Cell): List[Cell] = {
    if (start == target) return List(start)

    val queue = mutable.Queue[Int]()
    val visited = Array.ofDim[Boolean](height, width)
    val pathNodes = mutable.ArrayBuffer.empty[PathNode]

    visited(start.y)(start.x) = true
    pathNodes += PathNode(start, -1)
    queue.enqueue(0)

    while (queue.nonEmpty) {
      val pathId = queue.dequeue()
      val node = pathNodes(pathId)

      node.cell.neighbors.foreach { neighbor =>
        if (neighbor == target) {
          var cur = pathId
          var result = List(target)
          while (cur >= 0) {
            result = pathNodes(cur).cell :: result
            cur = pathNodes(cur).parent
          }
          return result.reverse
        }

        if (neighbor.isWalkable && !visited(neighbor.y)(neighbor.x)) {
          visited(neighbor.y)(neighbor.x) = true
          pathNodes += PathNode(neighbor, pathId)
          queue.enqueue(pathNodes.size - 1)
        }
      }
    }
    Nil
  }

  def midCellChecksum(path: List[Cell]): Long = {
    if (path.isEmpty) return 0L
    val cell = path(path.size / 2)
    ((cell.x * cell.y).toLong) & 0xffffffffL
  }

  override def run(iterationId: Int): Unit = {
    path = bfs(maze.start, maze.finish)
    resultVal = (resultVal + path.size) & 0xffffffffL
  }

  override def checksum(): Long = (resultVal + midCellChecksum(path)) & 0xffffffffL
}

class MazeAStar extends Benchmark {
  import MazeTypes._

  case class Item(priority: Int, vertex: Int) extends Ordered[Item] {
    override def compare(that: Item): Int = {
      if (priority != that.priority) priority.compare(that.priority)
      else vertex.compare(that.vertex)
    }
  }

  private var resultVal: Long = 0L
  private val width: Int = configVal("w").toInt
  private val height: Int = configVal("h").toInt
  private var maze: Maze = _
  private var path: List[Cell] = Nil

  override def name(): String = "Maze::AStar"

  override def prepare(): Unit = {
    maze = new Maze(width, height)
    maze.generate()
    resultVal = 0L
    path = Nil
  }

  def heuristic(a: Cell, b: Cell): Int =
    math.abs(a.x - b.x) + math.abs(a.y - b.y)

  def idx(y: Int, x: Int): Int = y * width + x

  def astar(start: Cell, target: Cell): List[Cell] = {
    if (start == target) return List(start)

    val size = width * height
    val cameFrom = Array.fill(size)(-1)
    val gScore = Array.fill(size)(Int.MaxValue)
    val bestF = Array.fill(size)(Int.MaxValue)

    val startIdx = idx(start.y, start.x)
    val targetIdx = idx(target.y, target.x)

    val openSet = mutable.PriorityQueue.empty[Item](Ordering[Item].reverse)
    val inOpen = new Array[Boolean](size)

    gScore(startIdx) = 0
    val fStart = heuristic(start, target)
    openSet.enqueue(Item(fStart, startIdx))
    bestF(startIdx) = fStart
    inOpen(startIdx) = true

    while (openSet.nonEmpty) {
      val current = openSet.dequeue()
      val currentIdx = current.vertex
      inOpen(currentIdx) = false

      if (currentIdx == targetIdx) {
        var cur = currentIdx
        var result = List.empty[Cell]
        while (cur != -1) {
          val y = cur / width
          val x = cur % width
          result = maze.cells(y)(x) :: result
          cur = cameFrom(cur)
        }
        return result.reverse
      }

      val currentY = currentIdx / width
      val currentX = currentIdx % width
      val currentCell = maze.cells(currentY)(currentX)
      val currentG = gScore(currentIdx)

      currentCell.neighbors.foreach { neighbor =>
        if (neighbor.isWalkable) {
          val neighborIdx = idx(neighbor.y, neighbor.x)
          val tentativeG = currentG + 1

          if (tentativeG < gScore(neighborIdx)) {
            cameFrom(neighborIdx) = currentIdx
            gScore(neighborIdx) = tentativeG
            val fNew = tentativeG + heuristic(neighbor, target)

            if (fNew < bestF(neighborIdx)) {
              bestF(neighborIdx) = fNew
              openSet.enqueue(Item(fNew, neighborIdx))
              inOpen(neighborIdx) = true
            }
          }
        }
      }
    }
    Nil
  }

  def midCellChecksum(path: List[Cell]): Long = {
    if (path.isEmpty) return 0L
    val cell = path(path.size / 2)
    ((cell.x * cell.y).toLong) & 0xffffffffL
  }

  override def run(iterationId: Int): Unit = {
    path = astar(maze.start, maze.finish)
    resultVal = (resultVal + path.size) & 0xffffffffL
  }

  override def checksum(): Long = (resultVal + midCellChecksum(path)) & 0xffffffffL
}
